src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -468,4 +501,273 @@ UA_PubSubKeyStorage_removeKeyStorage(UA_Server *server, UA_PubSubKeyStorage *key
+        /* call user callback to notify about the status */
+        sksClientContext *ctx = (sksClientContext *)client->config.clientContext;
+        UA_PubSubKeyStorage *ks = ctx->ks;
+        if(ks->sksConfig.sksPullRequestCallback)
+            ks->sksConfig.sksPullRequestCallback(ctx->server, connectStatus,
+                                            ks->sksConfig.userData);
+        addDelayedSksClientCleanupCb(client);
+    }
+    return;
+}
+
+static void
+setExternalEvenloopOnSkSClient(UA_ClientConfig *cc, UA_EventLoop *externalEventloop) {
+    UA_assert(externalEventloop != NULL);
+
+    /* Stop and delete the client EventLoop */
+    UA_EventLoop *el = cc->eventLoop;
+    if(el && !cc->externalEventLoop) {
+        if(el->state != UA_EVENTLOOPSTATE_FRESH &&
+           el->state != UA_EVENTLOOPSTATE_STOPPED) {
+            el->stop(el);
+            while(el->state != UA_EVENTLOOPSTATE_STOPPED) {
+                el->run(el, 100);
+            }
+        }
+        el->free(el);
+        cc->eventLoop = NULL;
+    }
+    cc->eventLoop = externalEventloop;
+    cc->externalEventLoop = true;
+}
+
+UA_StatusCode
+getSecurityKeysAndStoreFetchedKeys(UA_Server *server, UA_PubSubKeyStorage *keyStorage) {
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+    UA_UInt32 startingTokenId = UA_REQ_CURRENT_TOKEN;
+    UA_UInt32 requestKeyCount = UA_UINT32_MAX;
+
+    UA_ClientConfig *cc = keyStorage->sksConfig.sksClientConfig;
+    cc->stateCallback = onConnect;
+
+    setExternalEvenloopOnSkSClient(cc, server->config.eventLoop);
+
+    /* this is cleanedup in sksClientCleanupCb */
+    sksClientContext *ctx   = (sksClientContext *)UA_calloc(1, sizeof(sksClientContext));
+    if(!ctx)
+         return UA_STATUSCODE_BADOUTOFMEMORY;
+    ctx->ks = keyStorage;
+    ctx->server = server;
+    ctx->startingTokenId = startingTokenId;
+    ctx->requestedKeyCount = requestKeyCount;
+    cc->clientContext = ctx;
+
+    UA_Client *client = UA_Client_newWithConfig(cc);
+    if(!client)
+         return retval;
+
+    /* connect to sks server */
+    retval = UA_Client_connectAsync(client, keyStorage->sksConfig.sksEndpointUrl);
+    if(retval != UA_STATUSCODE_GOOD) {
+         UA_LOG_ERROR(&client->config.logger, UA_LOGCATEGORY_CLIENT,
+                      "Failed to connect SKS server with error: %s ",
+                      UA_StatusCode_name(retval));
+         /* we want to delete client without deleting config, otherwise pointer to
+          * eventloop will be also lost forever and ever */
+         /* Delete the async service calls with BADHSHUTDOWN */
+         __Client_AsyncService_removeAll(client, UA_STATUSCODE_BADSHUTDOWN);
+         /* Remove the internal regular callback */
+         UA_Client_removeCallback(client, client->houseKeepingCallbackId);
+
+         client->houseKeepingCallbackId = 0;
+
+         UA_free(client->config.clientContext);
+         /* the config will be reused for next call */
+         UA_free(client);
```

Comment:
Use UA_Client_delete.
This might work now. But in the future the client may carry additional dynamically allocated members.
So UA_Client_delete is the only "official" way of cleaning up and deleting the client.

But that also cleans up the client config.

Add UA_ClientConfig_copy (and taking a const config pointer for the SKS config) should be straightforward though.

Reply:
UA_ClientConfig_copy should do a deep copy otherwise, delete client removes references to logger, eventloop and other pointers.  It requires more works then I estimated, shallow copy will not work.

Reply:
The only problematic part is the handling of security policies.
Those are stateful when added to the configuration.

you could special-case them and reset those fields to NULL before cleaning up.

a generic configuration copy would be even better of course.

Reply:
> a generic configuration copy would be even better of course.

I agree. A suggestion I would like to make is to add "copy" function pointer to the all the plugin interface for security policy, eventloop etc. They all have void pointers which are casted to internal structs and make it hard to deep copy onto other location. We already have clear function pointer, so having copy will help.